Added Demographics Uploading Capability to AppCore

PacificCoaster opened the pull request:
Basic Demographic Data includes Patient Age, Wake Time and Sleep Time, Biological Sex, Height and Weight.

Created a new  APCDemographicUploader  class to access user data and upload the data.

Added a call to  APCAllSetContentViewController to upload the demographic data after initial sign up.

Added code to  APCProfileViewController  to upload demographic data when edits are made.

Added code to  APCAppDelegate  to check if there has been a previous demographic upload, and to perform the upload if it has never been performed before.

Tested by a complete new install, and verified that data was uploaded in the APCAllSetContentViewController.

Tested by a subsequent App install (upgrade), and verified that data was uploaded in the APCAppDelegate

Tested by making changes to data via the editing functions of APCProfileViewController, and verified that data was uploaded at the point where the patient switched away from APCProfileViewController.

Commit:
[APPLE-2873] - Demographic data (collected during onboarding) is not uploaded to Bridge

Created a new  APCDemographicUploader  class to access user data and upload the data.

Added a call to  APCAllSetContentViewController  to upload the demographic data after initial sign up.

Added code to  APCProfileViewController  to upload demographic data when edits are made (somewhat blunt instrument right now)

Added code to  APCAppDelegate  to check if there has been a previous demographic upload, and to perform the upload if it has never been performed before.

Signed-off-by: Henry McGilton <henry.mcgilton@ymedialabs.com>

# Conflicts:
#	APCAppCore/APCAppCore/Startup/APCAppDelegate.m
#	APCAppCore/APCAppCore/UI/Onboarding/SignUp/APCAllSetContentViewController.m
#	APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCProfileViewController.m

Commit:
[APPLE-2873] - Demographic data (collected during onboarding) is not uploaded to Bridge

Added Check for User Consented in APCAppDelegate, so that the user demographics are collected at end of Sign-Up, and only after verification.

Signed-off-by: Henry McGilton <henry.mcgilton@ymedialabs.com>

Commit:
[APPLE-2873] - Demographic data (collected during onboarding) is not uploaded to Bridge

Changed uploading of user's date of birth to user's age.

Added an  ageFromDateOfBirth  to the NSDate+Helper category to compute age given a date.

Changed  APCDemographicUploader  to upload the user's age instead of date of birth.  Changed the kPatientBirthDateKey to kPatientCurrentAgeKey

Signed-off-by: Henry McGilton <henry.mcgilton@ymedialabs.com>

Commit:
Demographic Data Uploader Capability

Commit:
Demographic Data Uploader Capability

Signed-off-by: Henry McGilton <henry.mcgilton@ymedialabs.com>

Commit:
Merge branch 'master' of https://github.com/PacificCoaster/AppCore into github-master

Commit:
Remove redundant property

Commit:
Merge branch 'upstream-master' into github-master

# Conflicts:
#	APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCProfileViewController.m

jwe-apple commented:
Under review, thanks.

jwe-apple commented on APCAppCore/APCAppCore/Library/Categories/NSDate+Helper.m:
```diff
@@ -134,6 +134,17 @@ - (NSString *) toStringInISO8601Format
+    NSUInteger  answer = 0;
+    if (dateOfBirth != nil) {
+        NSDateComponents  *differences = [[NSCalendar currentCalendar] components:(NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay)
+                                                                     fromDate:dateOfBirth toDate:[NSDate date] options:0];
```
If you're going to have a line break, please vertically align all the arguments

jwe-apple commented on APCAppCore/APCAppCore/Startup/APCAppDelegate.m:
```diff
@@ -121,6 +123,37 @@ - (BOOL)               application: (UIApplication *) __unused application
+        //    we run this code iff the user has previously consented,
+        //    indicating that this is an update to a previously installed version of the application
+        //
+    APCUser  *user = ((APCAppDelegate *)[UIApplication sharedApplication].delegate).dataSubstrate.currentUser;
```
This line is packing a lot into one. Not to mention it seems really ugly to create a dependency to the App Delegate to get the current user. Could this class be updated to take the current user for its initialization?

jwe-apple commented on APCAppCore/APCAppCore/Startup/APCAppDelegate.m:
```diff
@@ -121,6 +123,37 @@ - (BOOL)               application: (UIApplication *) __unused application
+            self.demographicUploader = [[APCDemographicUploader alloc] init];
+            [defaults setBool:YES forKey:kDemographicDataWasUploadedKey];
+            [defaults synchronize];
+            [self.demographicUploader uploadNonIdentifiableDemographicData];
```
Does “-uploadNonIdentifiableDemographicData” block? If so, that should be moved to a background queue.

PacificCoaster replied:
> On May 15, 2015, at 12:01 PM, jwe-apple notifications@github.com wrote:
> 
> In APCAppCore/APCAppCore/Startup/APCAppDelegate.m https://github.com/ResearchKit/AppCore/pull/59#discussion_r30436130:
> 
> > -        //    otherwise, the value should have been set to YES, to
> > -        //    indicate that the Demographic data was previously uploaded
> > -        //
> > -   
> > -        //
> > -        //    we run this code iff the user has previously consented,
> > -        //    indicating that this is an update to a previously installed version of the application
> > -        //
> > -    APCUser  *user = ((APCAppDelegate *)[UIApplication sharedApplication].delegate).dataSubstrate.currentUser;
> > -    if (user.isConsented) {
> > -        BOOL  demographicDataWasUploaded = [defaults boolForKey:kDemographicDataWasUploadedKey];
> > -        if (demographicDataWasUploaded == NO) {
> > -            self.demographicUploader = [[APCDemographicUploader alloc] init];
> > -            [defaults setBool:YES forKey:kDemographicDataWasUploadedKey];
> > -            [defaults synchronize];
> > -            [self.demographicUploader uploadNonIdentifiableDemographicData];
> >   Does “-uploadNonIdentifiableDemographicData” block? If so, that should be moved to a background queue.

I (think I) have addressed all of your comments.  Running some verification checks now.

The DemographicData uploader ends up by calling the  APCDataArchiverAndUploader uploadDictionary:demographics
The DataArchiverAndUploader does the uploading on a background queue.

```
. . . . . . . .    Henry
```

jwe-apple commented on APCAppCore/APCAppCore/UI/Onboarding/SignUp/APCAllSetContentViewController.m:
```diff
@@ -35,7 +35,7 @@
 #import "APCAllSetTableViewCell.h"
 
-static NSString *kAllSetCellIdentifier = @"AllSetCell";
+static  NSString  *kAllSetCellIdentifier = @"AllSetCell";
```
Added spaces should be removed.

jwe-apple commented on APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCDemographicUploader.m:
```diff
@@ -0,0 +1,125 @@
+
+- (void)uploadNonIdentifiableDemographicData
+{
+    APCUser  *user = ((APCAppDelegate *)[UIApplication sharedApplication].delegate).dataSubstrate.currentUser;
```
This method should take the user to update demographic data for instead of relying on getting it from the app delegate.

jwe-apple commented on APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCDemographicUploader.m:
```diff
@@ -0,0 +1,125 @@
+    if ((klarse == FP_INFINITE) || (klass == FP_NAN) || (klass == FP_ZERO)) {
+        [demographics setObject:@(0) forKey:kPatientWeightPoundsKey];
+    } else {
+        [demographics setObject:@(weightInPounds) forKey:kPatientWeightPoundsKey];
```
I think modern dictionary access would look nicer throughout this method: 

demographics[key] = <bool check> ? value : [NSNull null];

jwe-apple commented on APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCProfileViewController.m:
```diff
@@ -86,6 +86,9 @@ @interface APCProfileViewController () <ORKTaskViewControllerDelegate>
 @property (weak, nonatomic) IBOutlet UILabel *participationLabel;
 
+@property (strong, nonatomic) APCDemographicUploader  *demographicUploader;
+@property (nonatomic, assign)  BOOL                    profileEditsWerePerformed;
```
Too many spaces before BOOL

Commit:
Changes to AppCore Pull Request #59

Commit:
Changes to AppCore APCAppDelegate for Pull Request #59

Commit:
Changes to AppCore APCAllSetContentViewController for Pull Request #59

Commit:
Change to AppCore APCDemographicUploader Interface to Modify Initialiser for Pull Request #59

Commit:
Changes to AppCore APCDemographicUploader For Pull Request #59

Change designated initialiser to accept APCUser as parameter.

Declared APCUser as property in class continuation interface.

Modifications to code to use modern dictionary accessors.

Commit:
Changes to AppCore APCProfileViewController for Pull Request #59

Commit:
Merge branch 'upstream-master' into github-master

Commit:
Merge branch 'github-staging' into github-master

# Conflicts:
#	APCAppCore/APCAppCore/Library/Categories/NSDate+Helper.m
#	APCAppCore/APCAppCore/Startup/APCAppDelegate.m
#	APCAppCore/APCAppCore/UI/Onboarding/SignUp/APCAllSetContentViewController.m
#	APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCDemographicUploader.m
#	APCAppCore/APCAppCore/UI/TabBarControllers/Profile/APCProfileViewController.m
